src/backend.md

```diff
@@ -18,6 +19,236 @@
+| ------------------------- | ---------------------------------- |
+| `user.controller.ts`      | controller handlers implementation |
+| `user.controller.test.ts` | tests for controller handlers      |
+| `user.endpoints.ts`       | endpoints declaration              |
+| `handlers.ts`             |                                    |
+
+`handlers.ts` - special file which contains only 1 export instruction:
+
+```ts
+export * from './user.controller';
+```
+
+Controllers directory structure example:
+
+```
+- controllers
+  - user
+    - user.controller.ts
+    - user.controller.test.ts
+    - user.endpoints.ts
+    - handlers.ts
+  - endpoints.ts
+  - index.ts
+```
+
+For some resources `*.controller.test.ts` file can become huge (> 1500 rows). So if it exceeds some comfortable size for reading it may be splitted into separate files and put them into `tests` directory inside resource directory.
+
+Assume that user resource has following endpoint names: `getUsers`, `getUser`, `createUser`. So each test file should be named as `{endpoint name}.controller.test.ts`.
+
+Example:
+
+```
+- controllers
+  - user
+    - tests
+      - getUsers.controller.test.ts
+      - getUser.controller.test.ts
+      - createUser.controller.test.ts
+    - user.controller.ts
+    - user.endpoints.ts
+    - handlers.ts
+  - endpoints.ts
+  - index.ts
+```
+
+#### events directory
+
+Event handlers for resources should be placed into separated directories named as resource. For example: `user`, `role`, e.t.c.
+
+Events directory structure example:
+
+```
+- events
+  - user
+    - user.event.ts
+    - user.event.test.ts
+```
+
+#### jobs directory
+
+Job handlers for resources should be placed into separated directories named as resource. For example: `user`, `role`, e.t.c.
+
+Jobs directory structure example:
+
+```
+- jobs
+  - user
+    - user.job.ts
+    - user.job.test.ts
+  - jobNames.ts
+  - jobHandlers.ts
+```
+
+`jobName.ts` - file which contains enum with all existing job names.
+
+Example:
+
+```ts
+export enum JobName {
+  handleUserCreated = 'handleUserCreated',
+  handleUserUpdated = 'handleUserUpdated',
+}
+```
+
+`jobHandlers.ts` - file which contains mapping between job names and their handler functions.
+
+Example:
+
+```ts
+import { handleUserCreatedJob } from './user/handleUserCreated.job';
+import { handleUserUpdatedJob } from './user/handleUserUpdated.job';
+import { JobName } from './jobNames';
+
+export function getJobFunctions() {
+  return {
+    [JobName.handleUserCreated]: handleUserCreatedJob,
+    [JobName.handleUserUpdated]: handleUserUpdatedJob,
+  };
+}
+```
+
+#### lib directory
+
+This directory contains different helpers and utilities which are common for a whole service
+or used widely across multiple other services. These helpers should not contain any business logic
+rules or core related to service domain area. Directory may contain `*.ts` files of nested folders (only 1 level allowed here).
```

Comment:
So, `lib/xxx.ts` and `lib/xxx/yyy.ts` both are allowed, right?

Reply:
Another good question. 

I checked our new project `winston` and found that @dmnorc used nested folders inside `lib` directory.
I don't see anything criminal here because some helpers and utilities can be grouped inside `lib/{...}` folder

Reply:
> I don't see anything criminal here

agree

Reply:
so let's reformulate somehow to make it more clear